src/zkevm_specs/evm_circuit/execution/callop.py

```diff
@@ -147,7 +150,92 @@ def callop(instruction: Instruction):
             is_create=Transition.same(),
             code_hash=Transition.same_word(),
         )
-    else:
+    # precompiles call
+    elif is_precheck_ok and is_precompile == FQ.one():
+        precompile_return_length = instruction.curr.aux_data[0]
+        min_rd_copy_size = min(precompile_return_length, call.rd_length.n)
+
+        # precompiles have on code
+        instruction.constrain_equal(no_callee_code, FQ.one())
+        # precompiles address must be warm
+        instruction.constrain_equal(is_warm_access, FQ.one())
+
+        # Setup next call's context.
+        for field_tag, expected_value in [
+            (CallContextFieldTag.IsSuccess, call.is_success),
+            (CallContextFieldTag.CalleeAddress, callee_address_word),
+            (CallContextFieldTag.CallerId, instruction.curr.call_id),
+            (CallContextFieldTag.CallDataOffset, call.cd_offset),
+            (CallContextFieldTag.CallDataLength, call.cd_length),
+            (CallContextFieldTag.ReturnDataOffset, call.rd_offset),
+            (CallContextFieldTag.ReturnDataLength, call.rd_length),
+        ]:
+            instruction.constrain_equal_word(
+                instruction.call_context_lookup_word(field_tag, RW.Write),
+                WordOrValue(expected_value),
+            )
+
+        # Save caller's call state
+        for field_tag, expected_value in [
+            (CallContextFieldTag.ProgramCounter, instruction.curr.program_counter + 1),
+            (
+                CallContextFieldTag.StackPointer,
+                instruction.curr.stack_pointer + stack_pointer_delta,
+            ),
+            (CallContextFieldTag.GasLeft, instruction.curr.gas_left - gas_cost - callee_gas_left),
+            (CallContextFieldTag.MemorySize, call.next_memory_size),
+            (
+                CallContextFieldTag.ReversibleWriteCounter,
+                instruction.curr.reversible_write_counter + 1,
+            ),
+            (CallContextFieldTag.LastCalleeId, callee_call_id),
+            (CallContextFieldTag.LastCalleeReturnDataOffset, FQ.zero()),
+            (CallContextFieldTag.LastCalleeReturnDataLength, FQ(precompile_return_length)),
+        ]:
+            instruction.constrain_equal(
+                instruction.call_context_lookup(field_tag, RW.Write, callee_call_id),
+                expected_value,
+            )
+
+        # Verify data copy from precompiles
+        return_copy_rwc_inc = FQ.zero()
```

Comment:
Yep, I didn't forget this part and I added the copy of input_data and output_data. However, I still feel it's not right so I removed it.

My arguments, 
A. We didn't use any rlced data in our precompiles (e.g. `ecRecover`, `datacopy` ...etc). We use `word` type for our input, output and returned data directly in our precompiles.
B. If we add the copy of `input_data` and `output_data`, we should have some constraints for those rlced data (For example, constrain the equality of rlced data in `precompile_gadget` or in every precompile). But it doesn't make sense to me because we didn't use rlced data in our implementation. 

A question, do we need to constrain the equality of the data in `callop` and the data in individual precompile gadget? I guess the answer is yes. But how do we constrain the data consistency if it's non-fixed length? I don't have any solutions for now.

Conclusion,
1. keep the copy of input_data and output_data and add data consistency in `precompile_gadget` and every precompiles.
2. we don't do the copy since we didn't use rlc here but we need to figure out a way to constrain data consistency

Does it make sense to you?

Reply:
Don't we need to ensure that the output is properly copied to the return data buffer?

I think the only way to ensure the data consistency if it's not fixed length is to use RLC.

Reply:
I may have missed something, but I see no reason why the following doesn't work.

1. Copy with RLC to constrain the
  1.  Caller's memory for input === precompile input RLC
  2. Caller's memory for output === precompile output RLC
2. Parse the precopmile input RLC with words if the input is a word. We perform the same parsing for the precompile output RLC.

Reply:
@ChihChengLiang, it makes sense to me. Could you check [14a38d4](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/508/commits/14a38d42df1fd14fa7745949d4a4741c7f9f2cc1) to see if it fits your point 4?

Reply:
@KimiWu123 [b5816af](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/508/commits/b5816afe36977ad1b665848e5e47107f2254e5c8) The fix here is exactly what I was looking for.
The parsing in [f3eee6f](https://github.com/privacy-scaling-explorations/zkevm-specs/pull/508/commits/f3eee6f6f45b079af55251945d59e20a0d517b58) looks correct to me too!
Thanks for the fix!